registry/etcd/etcd.go

```diff
@@ -326,67 +373,122 @@ func (e *etcdRegistry) GetService(name string, opts ...registry.GetOption) ([]*r
 	ctx, cancel := context.WithTimeout(context.Background(), e.options.Timeout)
 	defer cancel()
 
-	rsp, err := e.client.Get(ctx, servicePath(name)+"/", clientv3.WithPrefix(), clientv3.WithSerializable())
-	if err != nil {
-		return nil, err
+	// parse the options and fallback to the default domain
+	var options registry.GetOptions
+	for _, o := range opts {
+		o(&options)
+	}
+	if len(options.Domain) == 0 {
+		options.Domain = defaultDomain
 	}
 
-	if len(rsp.Kvs) == 0 {
+	var results []*mvccpb.KeyValue
+	if options.Domain == registry.WildcardDomain {
+		rsp, err := e.client.Get(ctx, prefix, clientv3.WithPrefix(), clientv3.WithSerializable())
+		if err != nil {
+			return nil, err
+		}
+
+		// filter using a check for the service name
+		keyPath := fmt.Sprintf("/%v/", serializeServiceName(name))
+		for _, kv := range rsp.Kvs {
+			if strings.Contains(string(kv.Key), keyPath) {
```

Comment:
What does `kv.Key` look like? Do you have some examples? I'm always a bit worried with `strings.Contains` whether it might have a false positive. for example if `serializeServiceName(name)` returned `github.com/food/bar` and you were looking for something which was a substring there like `foo` you'd still get a hit

Reply:
kv.Key is something like `micro/registry/foo/go.micro.service.helloworld/1.0.0/nodeid`. I added / to either side of the service name so it'll only match exact service names. I spent some time thinking about this and I can't find a better solution.